### PR TITLE
Add PSR-7 erratum "Validation of Header Names and Values"

### PR DESCRIPTION
**Pull request by TimWolla**

### CODE REVIEW

**Review by Jean85 — CHANGES_REQUESTED** [review]
I'm not sure if this fits in the specification, I would defer this question to @weierophinney.

Also, I would suggest you to put this for discussion on our mailing list: https://groups.google.com/g/php-fig

**Comment by TimWolla** [comment]
> I'm not sure if this fits in the specification, I would defer this question

I've already shortly discussed / proposed that on the Discord with mwop generally being positive about this.

I've written this PR / proposal up yesterday evening to already have some basis to think and discuss about. If a proper mailing list discussion is desired or required then I'd be happy start one, but that would need to wait until April, so that I'm able to properly monitor and react to it.

**Comment by Jean85** [comment]
Good! This should follow the [Errata bylaw](https://www.php-fig.org/bylaws/psr-workflow/#erratas), so it requires the standard CC vote procedure (two weeks discussion, then two weeks for voting).

**Comment by weierophinney** [comment]
I think this is perfectly worded, honestly. 

As editor, it lies on me to open the discussion period and call for a vote; however, I also know that @TimWolla will be out-of-pocket sometime in the next few weeks. When would you like me to start the discussion period (i.e., when would be a time that you would be available to answer questions and respond to feedback)?

**Comment by TimWolla** [comment]
> When would you like me to start the discussion period (i.e., when would be a time that you would be available to answer questions and respond to feedback)?

I should be fully available again around April, 6th (give or take a few days for the usual after-vacation stuff). However I've already subscribed to the Google Group just now.

**Comment by weierophinney** [comment]
@TimWolla I'll start the discussion period on the 6th, then!

File: accepted/PSR-7-http-message-meta.md [hunk 1]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+A minimally viable validator is expected to reject header values containing the
+following characters:
+
+- NUL (0x00)
+- `\r` (0x0D)
+- `\n` (0x0A)
+
+If compatibility with older systems is desired then the sequence `\r\n` (0x0D0A)
+within a header value may be accepted, if and only if it is preceded by either
+SPACE (0x20) or `\t` (0x09).
```
Comment (TimWolla) [inline accepted/PSR-7-http-message-meta.md:687]:
```suggestion
If compatibility with older systems is desired then the sequence `\r\n` (0x0D0A)
within a header value may be accepted if and only if it is preceded by either
SPACE (0x20) or `\t` (0x09).
```
I don't think the comma is correct there, but I'm also not a native speaker.

File: accepted/PSR-7-http-message-meta.md [hunk 2]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+the parsing of the serialized message in a way that the contents of unrelated
+headers are changed. This misparsing can open up an application to security
+vulnerabilities. A common type of vulnerability is CRLF injection, allowing
+an attacker to inject additional headers or end the list of headers early.
+
+For this reason classes implementing the `MessageInterface` SHOULD strictly
+validate the header names and contents according to the most recent HTTP
+specification ([RFC 7230#3.2][1] at the time of writing). Incorrect values
+SHOULD be rejected and no attempt SHOULD be made to automatically correct
+the provided values.
```
Comment (TimWolla) [inline accepted/PSR-7-http-message-meta.md:665]: I'm not happy about the “no attempt SHOULD” and like to rephrase that to a real SHOULD NOT somehow.
```suggestion
specification ([RFC 7230#3.2][1] at the time of writing). The implementation
SHOULD reject invalid values and SHOULD NOT make any attempt to automatically correct
the provided values.
```

File: accepted/PSR-7-http-message-meta.md [hunk 3]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+
+## 7. Errata
+
+### 7.1 Validation of Header Names and Values
+
+Some special characters within the name or value of an HTTP header might affect
+the parsing of the serialized message in a way that the contents of unrelated
+headers are changed. This misparsing can open up an application to security
+vulnerabilities. A common type of vulnerability is CRLF injection, allowing
+an attacker to inject additional headers or end the list of headers early.
```
Comment (TimWolla) [inline accepted/PSR-7-http-message-meta.md:659]:
```suggestion
an attacker to inject additional headers or to end the list of headers early.
```

**Review by TimWolla — COMMENTED** [review]
Some small adjustments I've noticed while reading through this once more.

File: accepted/PSR-7-http-message-meta.md (hunk 1 above)
Comment (weierophinney) [reply accepted/PSR-7-http-message-meta.md:687]: Comma definitely not necessary!

File: accepted/PSR-7-http-message-meta.md (hunk 2 above)
Comment (weierophinney) [reply accepted/PSR-7-http-message-meta.md:665]: The SHOULD NOT verbiage reads more clearly; nice catch!

File: accepted/PSR-7-http-message-meta.md [hunk 4]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+to the HTTP specification.
+
+A minimally viable validator is expected to reject header values containing the
+following characters:
+
+- NUL (0x00)
+- `\r` (0x0D)
+- `\n` (0x0A)
+
+If compatibility with older systems is desired then the sequence `\r\n` (0x0D0A)
```
Comment (GrahamCampbell) [inline accepted/PSR-7-http-message-meta.md:685]: Doesn't this contradict the earlier requirement to always block `\r`?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: accepted/PSR-7-http-message-meta.md [hunk 5]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+within a header value may be accepted if and only if it is preceded by either
+SPACE (0x20) or `\t` (0x09).
```
Comment (TimWolla) [inline accepted/PSR-7-http-message-meta.md:687]:
```suggestion
SPACE (0x20) or `\t` (0x09). The full sequence SHOULD then be internally normalized to a single SPACE (0x20).
```
This should ensure that the *output* of the implementation is always safe to pass into an implementation that rejects line folding.

File: accepted/PSR-7-http-message-meta.md [hunk 6]
```diff
@@ -647,3 +647,46 @@ used to populate the headers of an HTTP message.
...
+
+A minimally viable validator is expected to reject header values containing the
+following characters:
+
+- NUL (0x00)
+- `\r` (0x0D)
+- `\n` (0x0A)
+
+If compatibility with older systems is desired then the sequence `\r\n` (0x0D0A)
+within a header value may be accepted if and only if it is preceded by either
```
[the remark this answers is not in the document]
Comment (weierophinney) [reply accepted/PSR-7-http-message-meta.md:686]: :+1:

File: accepted/PSR-7-http-message-meta.md (hunk 5 above)
Comment (weierophinney) [reply accepted/PSR-7-http-message-meta.md:687]: :+1:

File: accepted/PSR-7-http-message-meta.md (hunk 4 above)
Comment (TimWolla) [reply accepted/PSR-7-http-message-meta.md:685]: Note: The above comment is factually incorrect. I've mixed up "preceded" and "followed": https://github.com/php-fig/fig-standards/pull/1274/commits/8bd62b21f6d477a57488165d2e0253206d848a00.

Correct is "CR LF SPACE", not "SPACE CR LF".

**Comment by weierophinney** [comment]
Errata vote was accepted: https://groups.google.com/g/php-fig/c/O3flCAGhurc/m/yPbGNAAuCQAJ

**Pull request merged**